### PR TITLE
[SYCL-MLIR]: Version affine loop using scf::IfOp

### PR DESCRIPTION
**Pull request by etiotto**
Affine loops can be versioned using a `non-affine` SCF if operation because the versioning condition might not always be a linear combination of the loop induction variable. 

This PR extends and simplify the loop versioning tools to allow versioning an Affine loop using either an `AffineIfOp` or an `scf::SCFIfOp`.

### CODE REVIEW

**Comment by etiotto** [comment]
Depends on #9026

**Comment by etiotto** [comment]
The failure in E2E test is not a regression caused by this PR so merging.

File: polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h [hunk 1]
```diff
@@ -40,16 +43,6 @@ void fully2ComposeAffineMapAndOperands(PatternRewriter &rewriter,
                                        DominanceInfo &DI);
 bool isValidIndex(Value val);
 
-//===----------------------------------------------------------------------===//
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h:43]: Removed - no longer needed

File: polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h [hunk 2]
```diff
@@ -70,76 +63,59 @@ class LoopVersionCondition {
...
-  LoopVersionBuilder(LoopLikeOpInterface loop) : loop(loop) {}
-
-  void versionLoop(RegionBranchOpInterface) const;
+  void createElseBody(scf::IfOp) const;
+  void createElseBody(AffineIfOp) const;
 
   mutable LoopLikeOpInterface loop; /// The loop to version.
 };
 
-/// Concrete class to version a SCF loop operation.
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h:103]: Removed derived classes - now `LoopVersionBuilder` can version either SCF loops or affine loops using the `LoopVersionCondition` supplied.

File: polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir [hunk 3]
```diff
@@ -111,6 +107,94 @@ func.func private @test(%arg0: memref<?x!sycl_accessor_1_i32_rw_gb, 4>, %arg1: m
...
 
+!sycl_id_1_ = !sycl.id<[1], (!sycl.array<[1], (memref<1xi64>)>)>
+!sycl_range_1_ = !sycl.range<[1], (!sycl.array<[1], (memref<1xi64>)>)>
+!sycl_accessor_impl_device_1_ = !sycl.accessor_impl_device<[1], (!sycl_id_1_, !sycl_range_1_, !sycl_range_1_)>
+!sycl_accessor_1_i32_rw_gb = !sycl.accessor<[1, i32, read_write, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xi32, 1>)>)>
+!sycl_accessor_1_i32_r_gb = !sycl.accessor<[1, i32, read, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xi32, 1>)>)>
+
+// CHECK: [[GUARD_COND:#.*]] = affine_set<() : (7 >= 0)>
+
+// COM: There is only one loop, i.e., the loop is not versioned.
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir:118]: remove

File: polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir [hunk 4]
```diff
@@ -111,6 +107,94 @@ func.func private @test(%arg0: memref<?x!sycl_accessor_1_i32_rw_gb, 4>, %arg1: m
...
+// CHECK-LABEL: testAffineLoopVersioning
```
Comment (whitneywhtsang) [inline polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir:119]: Should we simplify this test like the test below, and rely on the first test to test the codegen?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 5]
```diff
@@ -369,20 +366,26 @@ class VersionConditionBuilder {
   using SCFCondition = LoopVersionCondition::SCFCondition;
   using AffineCondition = LoopVersionCondition::AffineCondition;
 
-  std::unique_ptr<LoopVersionCondition> createCondition() const {
+  /// The kind of condition to create.
+  enum class ConditionKind { SCF, Affine };
+
+  std::unique_ptr<LoopVersionCondition>
+  createCondition(ConditionKind condKind) const {
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:373]: As `createConditionForAffineLoop` is actually not implemented, and we don't plan to for the near future, should we just remove `createConditionForAffineLoop` , rename `createConditionForSCFLoop` to `createCondition`, and simply always use `createCondition`?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1]: can you please change `willVersion` to allow affine loop?

Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1]: nvm, my change is not in.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:373]: Yes I am fine with that actually.

File: polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir (hunk 4 above)
Comment (etiotto) [reply polygeist/test/polygeist-opt/sycl/licm-accessor-versioning.mlir:119]: Sure

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]

**Pull request merged**